mmcv/ops/csrc/common/pytorch_device_registry.hpp

```diff
@@ -0,0 +1,135 @@
+#ifndef PYTORCH_DEVICE_REGISTRY_H
+#define PYTORCH_DEVICE_REGISTRY_H
+
+#include <torch/extension.h>
+
+#include <cassert>
+#include <functional>
+#include <map>
+#include <type_traits>
+
+inline std::string GetDeviceStr(const at::Device& device) {
+  std::string str = DeviceTypeName(device.type(), true);
+  if (device.has_index()) {
+    str.push_back(':');
+    str.append(std::to_string(device.index()));
+  }
+  return str;
+}
+
+// Registry
+template <typename F, F f>
+class DeviceRegistry;
+
+template <typename Ret, typename... Args, Ret (*f)(Args...)>
+class DeviceRegistry<Ret (*)(Args...), f> {
+ public:
+  using FunctionType = Ret (*)(Args...);
+
+  void Register(at::DeviceType device, FunctionType function) {
+    entries_.insert({device, function});
+  }
+
+  FunctionType Find(at::DeviceType device) const {
+    auto it = entries_.find(device);
+    if (it != entries_.end()) {
+      return it->second;
+    }
+    return nullptr;
+  }
+
+  static DeviceRegistry& instance() {
+    static DeviceRegistry inst;
+    return inst;
+  }
+
+ private:
+  DeviceRegistry() = default;
+  std::map<at::DeviceType, FunctionType> entries_;
+};
+
+// get device of first tensor param
+
+template <typename T, typename... Args,
+          std::enable_if_t<std::is_same<std::decay_t<T>, at::Tensor>::value,
+                           bool> = true>
+at::Device GetFirstTensorDevice(T&& t, Args&&... args) {
+  return std::forward<T>(t).device();
+}
+template <typename T, typename... Args,
+          std::enable_if_t<!std::is_same<std::decay_t<T>, at::Tensor>::value,
+                           bool> = true>
+at::Device GetFirstTensorDevice(T&& t, Args&&... args) {
+  return GetFirstTensorDevice(std::forward<Args>(args)...);
+}
+
+// check device consistency
+
+inline std::pair<int, at::Device> CheckDeviceConsistency(
+    const at::Device& device0, int index) {
+  return {index, device0};
+}
+
+template <typename T, typename... Args,
+          std::enable_if_t<!std::is_same<std::decay_t<T>, at::Tensor>::value,
```

Comment:
Is this feature supported since C++14? What is newest C++ version supported in mmcv?

Reply:
C++14 is required.
MMCV require gcc>=5.4, which can support all features in C++14

Reply:
actually if you are not comfortable adding the C++14 requirement, you can replace the `enable_if_t`  variant with `typename enable_if<...>::type`, which only requires C++11